### PR TITLE
Allow `ColumnInterface` as column type

### PR DESCRIPTION
**Pull request by Tigrov**
First step to solve yiisoft/db-migration#218, yiisoft/db-migration#219 and other

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ✔️
| Fixed issues  | -

### CODE REVIEW

File: src/QueryBuilder/AbstractDDLQueryBuilder.php [hunk 1]
```diff
@@ -169,14 +169,15 @@ public function createTable(string $table, array $columns, string $options = nul
     {
         $cols = [];
 
-        /** @psalm-var string[] $columns */
+        /** @psalm-var array<string, ColumnInterface>|string[] $columns */
```
Comment (vjik) [inline src/QueryBuilder/AbstractDDLQueryBuilder.php:172]:
```suggestion
        /** @psalm-var array<array-key, ColumnInterface|string> $columns */
```

File: src/QueryBuilder/AbstractDDLQueryBuilder.php [hunk 2]
```diff
@@ -169,14 +169,15 @@ public function createTable(string $table, array $columns, string $options = nul
...
         foreach ($columns as $name => $type) {
             if (is_string($name)) {
                 $cols[] = "\t"
                     . $this->quoter->quoteColumnName($name)
                     . ' '
                     . $this->queryBuilder->getColumnType($type);
             } else {
+                /** @psalm-var string $type */
                 $cols[] = "\t" . $type;
```
Comment (vjik) [inline src/QueryBuilder/AbstractDDLQueryBuilder.php:181]:
```suggestion
                $cols[] = "\t" . $type;
```

if `$name` is integer, then here will throw exception, becuase `ColumnInterface` is not stringable.

File: tests/Db/Command/CommandTest.php [hunk 3]
```diff
@@ -39,17 +40,20 @@ public function testAddCheck(): void
...
+    /** @dataProvider \Yiisoft\Db\Tests\Provider\CommandProvider::columnTypes */
+    public function testAddColumn(ColumnInterface|string $type): void
     {
         $db = $this->getConnection();
 
         $command = $db->createCommand();
-        $sql = $command->addColumn('table', 'column', SchemaInterface::TYPE_INTEGER)->getSql();
+        $sql = $command->addColumn('table', 'column', $type)->getSql();
+
+        $type = $db->getQueryBuilder()->getColumnType($type);
```
Comment (vjik) [inline tests/Db/Command/CommandTest.php:51]: Suggest use another name for variable. $type here and $type in method parameters are confused =)

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/QueryBuilder/AbstractDDLQueryBuilder.php (hunk 1 above)
Comment (Tigrov) [reply src/QueryBuilder/AbstractDDLQueryBuilder.php:172]: If column is `ColumnInterface` it should have name as an array key => `array<string, ColumnInterface>`
If it is `string` then the array key can be `int|string` => `string[]`

File: src/QueryBuilder/AbstractDDLQueryBuilder.php (hunk 2 above)
Comment (Tigrov) [reply src/QueryBuilder/AbstractDDLQueryBuilder.php:181]: `ColumnInterface` can be only with string `$name`

File: src/QueryBuilder/AbstractDDLQueryBuilder.php (hunk 1 above)
Comment (vjik) [reply src/QueryBuilder/AbstractDDLQueryBuilder.php:172]: Can you add columns format to phpdoc in `DDLQueryBuilderInterface`?

**Review by vjik — APPROVED** [review]

**Review by darkdef — APPROVED** [review]

**Review by rustamwin — APPROVED** [review]

**Review by vjik — CHANGES_REQUESTED** [review]
Need line in changelog only

**Review by vjik — APPROVED** [review]

**Pull request merged**